src/Structures/Tree.php

```diff
@@ -45,15 +45,25 @@ public function tree($tree = null)
     {
         return $this->fluentlyGetOrSet('tree')
             ->getter(function ($tree) {
-                $key = "structure-{$this->handle()}-{$this->locale()}-".md5(json_encode($tree));
-
-                return Blink::once($key, function () use ($tree) {
+                return Blink::once($this->treeCacheKey(), function () use ($tree) {
                     return $this->structure()->validateTree($tree, $this->locale());
                 });
             })
             ->args(func_get_args());
     }
 
+    public function clearTreeCache()
+    {
+        Blink::forget($this->treeCacheKey());
+
+        return $this;
+    }
+
+    public function treeCacheKey()
+    {
+        return "structure-{$this->handle()}-{$this->locale()}-".md5(json_encode($this->tree));
+    }
+
```

Comment:
This is overlapping with some of my work on #5899 - maybe we should get our heads together @ryanmitchell ?

Reply:
Yeah it does... does this test pass on your PR? https://github.com/statamic/cms/blob/d84b3b1e58073e1c00ba67cd4bfa1039d7097259/tests/Data/Entries/EntryTest.php#L1814

Reply:
Unlikely, as my implementation is different to yours and I'm not clearing any tree caches... I found that I didn't need what it looks like you're trying to solve here because of the way the key is based on the tree contents itself, it basically 'self-clears'

Reply:
My comment was more specifically around extracting this cache key out to its own function tho

Reply:
I pulled it down and it does pass, so I'll close this PR in favour of yours if you can add the following test to your `Tests\Data\Entries\EntryTest.php`

```
    /** @test */
    public function it_clears_the_tree_cache_when_saving_an_entry()
    {
        $collection = (new Collection)->handle('pages')->routes('{parent_uri}/{slug}');
        $collection->save();

        $a = tap(Entry::make()->collection('pages')->id('1')->slug('one')->merge(['title' => 1]))->save();
        $b = tap(Entry::make()->collection('pages')->id('2')->slug('two')->merge(['title' => 2]))->save();
        $c = tap(Entry::make()->collection('pages')->id('3')->slug('three')->merge(['title' => 3]))->save();

        $collection->structureContents([
            'expects_root' => false,
        ])->save();

        $collection->structure()->in('en')->tree(
            [
                [
                    'entry' => '1',
                    'children' => [
                        ['entry' => '2'],
                    ],
                ],
                [
                    'entry' => '3',
                ],
            ]
        )->save();

        $this->assertEquals($a->url(), '/one');
        $this->assertEquals($b->url(), '/one/two');
        $this->assertEquals($c->url(), '/three');
    }
```

Reply:
Yeah I was just refactoring that code out... so looks like your PR solves this #5046 too.